core/amber/src/main/scala/edu/uci/ics/amber/engine/common/ambermessage/WorkflowMessage.scala

```diff
@@ -1,12 +1,42 @@
 package edu.uci.ics.amber.engine.common.ambermessage
 
 import edu.uci.ics.amber.engine.common.virtualidentity.ActorVirtualIdentity
+import edu.uci.ics.amber.engine.common.virtualidentity.util.CLIENT
 
-sealed trait WorkflowMessage extends Serializable {
-  val from: ActorVirtualIdentity
-  val sequenceNumber: Long
+import scala.util.hashing.MurmurHash3
+
+object ChannelEndpointID {
+  def apply(endpointWorker: ActorVirtualIdentity, isControlChannel: Boolean): ChannelEndpointID = {
+    new ChannelEndpointID(endpointWorker, isControlChannel)
+  }
+}
+
+class ChannelEndpointID(val endpointWorker: ActorVirtualIdentity, val isControlChannel: Boolean)
+    extends Serializable {
+  def canEqual(other: Any): Boolean = other.isInstanceOf[ChannelEndpointID]
+
+  override def equals(other: Any): Boolean =
+    other match {
+      case that: ChannelEndpointID =>
+        (that canEqual this) && endpointWorker == that.endpointWorker && isControlChannel == that.isControlChannel
```

Comment:
`that canEqual this` seems unnecessary since you already used pattern match to ensure `that` is a ChannelEndpointID?